lib/utils/job-utils/diag-tool.js

```diff
@@ -0,0 +1,276 @@
+
+    /**
+    * Wrap fs.createReadStream as Promise
+    * @param {String} filename: file to be read to stream
+    * @return {Promise}
+    */
+    function createReadStream(filename){
+        return new Promise(function(resolve, reject){
+        var stream;
+        function onError(err){
+            reject(err);
+        }
+
+        function onReadable(){
+            cleanup();
+            resolve(stream);
+        }
+
+        function cleanup(){
+            stream.removeListener('readable', onReadable);
+            stream.removeListener('error', onError);
+        }
+
+        stream = fs.createReadStream(filename);
+        stream.on('error', onError);
+        stream.on('readable', onReadable);
+      });
+    }
+
+    /**
+    * Diag upload image file API
+    * @param {String} imageFilePath: local file path for image to be uploaded
+    * @param {String} uploadApi: diag API used to upload file
+    * @return {Promise}
+    */
+    DiagToolFactory.prototype.uploadImageFile = function(imageFilePath, uploadApi){
+        var self = this;
+        var api = uploadApi || '/api/upload/folder';
+
+        var options = {
+            uri: self.server + api,
+            method: 'POST'
+        };
+        return createReadStream(imageFilePath)
+        .then(function(stream){
+            options.formData = {file: stream};
+        })
+        .then(function(){
+            return request(options);
+        });
+    };
+
+    /**
+    * Diag discovery API
+    * @return {Promise}
+    */
+    DiagToolFactory.prototype.runSyncDiscovery = function(){
+        var self = this;
+        var api = '/api/system/tests/discovery/sync/run';
+        var options = {
+            uri: self.server + api,
+            method: 'GET',
+            json: true
+        };
+        return request(options);
+    };
+
+    /**
+    * Diag discovery API with retry
+    * @param {integer} delay: time interval for each retry
+    * @param {integer} retries: retry count
+    * @return {Promise}
+    */
+    DiagToolFactory.prototype.retrySyncDiscovery = function(delay, retries){
+        var self = this;
+        return self.runSyncDiscovery()
+        .catch(function(err){
+            if (_.has(err, 'statusCode') && err.statusCode !== 500) {
+                logger.error('Faied to run diag discovery', {
+                    error: err,
+                    nodeId: self.nodeId
+                });
+                throw err;
+            } else if (retries > 0) {
+                retries = retries - 1;
+                return Promise.delay(delay)
+                .then(function(){
+                    return self.retrySyncDiscovery(delay*2, retries);
+                });
+            } else {
+                logger.error('Faied to run diag discovery', {
+                    error: 'Timeout',
+                    nodeId: self.nodeId
+                });
+                throw new Error('Failed to connect to diag on node, timeout');
```

Comment:
Is it necessary to print log here as well as throw an error? I assume the caller layer will print this rejection somewhere. Same question for https://github.com/RackHD/on-tasks/pull/471/files#diff-5319045c236237bd82248d89042dcf2eR204